docs/madr/decisions/081-datasource-api.md

```diff
@@ -0,0 +1,481 @@
+	Type           SecretDataSourceType `json:"type,omitempty"`
+	File           *File                `json:"file,omitempty"`
+	SecretRef      *SecretRef           `json:"secretRef,omitempty"`
+	EnvVar         *EnvVar              `json:"envVar,omitempty"`
+	InsecureInline *Inline              `json:"insecureInline,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=File;EnvVar;Inline
+type DataSourceType string
+
+const (
+	DataSourceFile   DataSourceType = "File"
+	DataSourceEnvVar DataSourceType = "EnvVar"
+	DataSourceInline DataSourceType = "InsecureInline"
+)
+
+type DataSource struct {
+	Type   DataSourceType `json:"type,omitempty"`
+	File   *File          `json:"file,omitempty"`
+	EnvVar *EnvVar        `json:"envVar,omitempty"`
+	Inline *Inline        `json:"inline,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=ControlPlane;DataPlane
+// default: ControlPlane
+type ResolveOn string
+
+const (
+	ResolveOnCP ResolveOn = "ControlPlane"
+	ResolveOnDP ResolveOn = "DataPlane"
+)
+
+type File struct {
+	ResolveOn ResolveOn `json:"resolveOn,omitempty"`
+	Path      string    `json:"path,omitempty"`
+}
+
+type EnvVar struct {
+	ResolveOn ResolveOn `json:"resolveOn,omitempty"`
+	Name      string    `json:"name,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Secret
+type RefType string
+
+const (
+	SecretRefType RefType = "Secret"
+)
+
+type SecretRef struct {
+	Kind RefType `json:"kind,omitempty"`
+	Name string  `json:"name,omitempty"`
+}
+
+type Inline struct {
+	Value string `json:"value,omitempty"`
+}
+```
+
+The proposed model introduces clear semantics for resource resolution, allowing users to specify whether a file (or other data source) should be loaded by the control plane or the data plane. Additionally, it separates secret-related data sources from general ones, making it clearer which options are allowed in which contexts.
+
+#### Pros
+* Clear API boundaries - users can explicitly configure where the resource should be loaded.
+* Inline secrets are clearly marked as insecure, improving visibility and security awareness.
+
+#### Cons
+* Two separate models, which may introduce some code duplication and increase maintenance overhead.
+
+### Option 4: Separate Structures for Secrets and Other Data Sources with an Resolution Context directly in the resource
+
+The idea is to avoid embedding the resolution logic (i.e., where the resource should be resolved) into the `DataSource` itself. Instead, this responsibility should remain with the resource that uses it. This way, we keep the `DataSource` object clean and focused.
+
+```golang
+// +kubebuilder:validation:Enum=File;Secret;EnvVar;InsecureInline
+type ControlPlaneDataSourceType string
+
+const (
+	CPDataSourceFile      ControlPlaneDataSourceType = "File"
+	CPDataSourceSecretRef ControlPlaneDataSourceType = "Secret"
+	CPDataSourceEnvVar    ControlPlaneDataSourceType = "EnvVar"
+	CPDataSourceInline    ControlPlaneDataSourceType = "InsecureInline"
+)
+
+type ControlPlaneDataSource struct {
```

Comment:
I'm not sure if this one should be called `ControlPlaneDataSource`. I like your previous suggestion `SecretDataSource` more. Because it's not really about CP or not CP, if user inlines private key it's considered insecure (even if DP reads it), but if they inline cert it's fine. 

Alternatively, we can call it either `SecureDataSource`, `ProtectedDataSource` or `SensitiveDataSource`